design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,276 @@
+    ```
+    {
+      "service_name": "mysql_service",
+      "service_type": "myql",
+      "start_command": "mysql_start.sh", # imagine that "mysql_start.sh" calls out to "helm install"
+      "stop_command": "mysql_stop.sh"
+    }
+    ```
+1. Service Configs exist for Conjur and MySQL. These files would exist as `./mysql_config.json` and `./conjur_config.json`. An example for Conjur is provided below
+    ```
+    {
+      "conjur_account": "test",
+      "version": "10.9",
+      "docker_image": "registry.tld/conjur-appliance:10.9-stable"
+      "admin_user": "kumbi",
+      "admin_password": "definetelyasecret"
+    }
+    ```
+1. A service instances file exists for the test case under consideration. This file contains an array of Service Definition and Service Config pairs. This files would be called `services.json`
+    ```
+    [
+      {
+          "config": "./mysql_config.json",
+          "service_definition": "mysql_service",
+      },
+      {
+          "config": "./conjur_config.json",
+          "service_definition": "conjur_service"
+      }
+    ]
+    ```
+1. There is a test case written in Gherkin for validating using the MySQL service connector with Conjur. See below that there is an expliit step specifying the services that are expected to be running. 
+    ```
+    Feature: Secretless happy path with database connectors
+    
+      Scenario: Connection to MySQL with MySQL credentials stored in Conjur
+        Given services are running as defined in "./services.json" and the results are store in context path "services"
+        When I store secrets defined in "./conjur-secrets.json" to Conjur using credentials in context path "services.conjur_service.creds"
+        And I create an identity for an application as defined in "./application_identity.json" using credentials in context path "services.conjur_service.creds""
+        And I deploy a test app as defined in "./test-app.yml"
+        Then the app is able to connect to MYSQL on "localhost:332"
+    ```
+1. The command `cucumber` is called to run test cases. The command is configurable and allows the runner to select a subset of all tests to run.
+
+## Interfaces
+
+### Opaque API for lifecycle management of dependency resources
+
+The API makes it possible to implement service lifecycle methods in whatever way makes sense. That could be literal bash, docker-compose, calling an HTTP API to grab the service from a pool etc. This makes it possible to have similar experience between different environments (Jenkins, developer machine etc.). The same exact commands would be called on Jenkins as on the developer machine. The action of the commands would depend on the environment. For example, the developer machine can use a less resource intensive implementation of the service and Jenkins can run something that more closely approximates customer environments.
```

Comment:
Given that Jenkins only has docker, we should try to strive for dockerized or bash impls